Orbax Loading and Sharding Support feature

amitsrivastava78 opened the pull request:
Support
1. New loading function for the orbax checkpointing
2. Support sharding feature

orbax-dev commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,49 @@
     h5py = None
 
 
+def _is_orbax_checkpoint(filepath):
```
Please don't look at internal files, it's a bit of a brittle behavior. 

You can just use this function: https://source.corp.google.com/piper///depot/google3/third_party/py/orbax/checkpoint/experimental/v1/__init__.py;l=27

amitsrivastava78 replied:
ok

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -167,6 +175,62 @@ def __init__(
+            return False
+
+        try:
+            import orbax.checkpoint as ocp
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -204,11 +268,13 @@ def _save_checkpoint(self, step, logs=None):
-        # save_pytree is called
-        if self.verbose > 0:
+        # Synchronize before saving to ensure consistency
+        self._sync_processes(f"checkpoint_save_start_{step}")
```
ok

hertschuh commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -297,3 +359,6 @@ def wait_until_finished(self):
                 time.sleep(0.1)
+
+        # Multi-host synchronization: ensure all hosts complete
+        self._sync_processes("checkpoint_wait_complete")
```
Just curious, if you're doing `_sync_processes`, do you still need the `wait` before?

amitsrivastava78 replied:
Yes This ensures that the local asynchronous checkpoint operations on the current host have actually completed. Without this, the host might proceed while its save operations are still running in the background._sync_processes this acts as a barrier synchronization across all hosts

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+                    flat_dict[f"{prefix}{key}"] = value
+
+        _flatten(nested_dict)
+        return flat_dict
```
This already exists, it's `tree.flatten_with_path` (`from keras.src import tree`), you'll get a list of tuple with path as a tuple and value.

So `_flatten_nested_dict` is equivalent to:
```python
{".".join(p):v for p, v in tree.flatten_with_path(nested_dict)}
```

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+            self.assertTrue(
+                np.allclose(trained_np, loaded_np),
+                f"Metrics variable {key} should match",
+            )
```
Please also remove `compare_nested_dicts` in the existing tests above.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+                layout_map["output_layer/kernel"] = TensorLayout(
+                    axes=(None, None)
+                )
+                layout_map["output_layer/bias"] = TensorLayout(axes=(None,))
```
But this is not proving anything if there is only 1 device.

amitsrivastava78 replied:
fixed the issue by adding a skip condition for single-device scenarios.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+
+                # Check that loading actually changed some weights
+                loading_changed_weights = any(
+                    not np.allclose(init, loaded)
```
Use `self.assertAllClose`

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+                    not np.allclose(init, loaded)
+                    for init, loaded in zip(initial_weights, loaded_weights)
+                )
+                self.assertTrue(
```
Don't use `assertTrue`, it won't tell you which one if off, by how much, etc.

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+        """Test multi-host checkpointing functionality (JAX only)."""
+        self._test_multihost_checkpointing()
+
+    def _test_multihost_checkpointing(self):
```
Why the private and public methods?

amitsrivastava78 replied:
simplified the test structure by removing the unnecessary private methods

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+
+        # Mock multi-host environment
+        with mock.patch("orbax.checkpoint.multihost") as mock_multihost:
+            # Test when multi-host is initialized
```
This test is very contrived, it mocks one thing and then checks a side effect that is internal and not user visible. I would just remove this.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+        """Test multi-host synchronization methods (JAX only)."""
+        self._test_multihost_synchronization_methods()
+
+    def _test_multihost_synchronization_methods(self):
```
Why the private and public methods?

amitsrivastava78 replied:
Removed as part of the removal of the test case

hertschuh reviewed (commented):
Thanks for the PR!

Can you tighten the `orbax_checkpoint_test.py` file. It's extremely long and hard to follow. I think:
- a lot fewer tests could cover basically the same
- some parameterized tests could minimize code duplication
- the verification blocks could be much shorter using `keras.tree` and `self.assertAllClose`, I gave some examples.

orbax-dev reviewed (commented):
LGTM, it won't let me approve with comments.

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+        self._test_model_load_with_saving_mode(save_on_background=True)
+
+    def _test_model_load_with_saving_mode(self, save_on_background):
+        """Helper method to test Model.load() with different saving modes."""
```
I merged 5 separate test methods into 1 parameterized test:

test_checkpoint_loading
test_checkpoint_loading_weights_only
test_checkpoint_loading_with_optimizer_state
test_checkpoint_loading_with_metrics_state
test_model_load_method (and its helper _test_model_load_with_saving_mode)
New Parameterized Test test_checkpoint_loading_comprehensive

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+        reason="Checkpoint structure tests require JAX backend",
+    )
+    def test_distributed_checkpoint_directory_structure(self):
+        """Test OrbaxCheckpoint directory structure for distributed training."""
```
Refactored the test to focus on Keras functionality rather than Orbax internals

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+            self.assertTrue(
+                callback.is_primary_host(),
+                "Should be primary host in mock setup",
+            )
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -577,3 +587,801 @@ def compare_nested_dicts(orig_dict, loaded_dict):
+            mock_multihost.is_primary_host.return_value = True
+            mock_multihost.sync_global_processes = mock.MagicMock()
+
+            callback._multihost_initialized = True
```
removed as part of removal of test case

amitsrivastava78 commented on keras/src/models/model.py:
```diff
@@ -424,6 +425,145 @@ def load_weights(self, filepath, skip_mismatch=False, **kwargs):
         )
 
+    @traceback_utils.filter_traceback
+    def load(self, filepath):
```
ok

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,44 @@
+        return ocp.is_orbax_checkpoint(filepath)
+    except (ImportError, AttributeError):
+        # Fallback to basic directory check if Orbax API not available
+        return file_utils.isdir(filepath)
```
updated the fallback to check for the existence of an orbax chk pt file

hertschuh commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -195,23 +260,18 @@ def _save_checkpoint(self, step, logs=None):
-                f"OrbaxCheckpoint: Triggering async save for step {step}..."
-            )
+            composite_state = {
+                "model_config": self.model.get_config(),
```
Hmm... there is also the issue of the assets: https://github.com/keras-team/keras/blob/master/keras/src/saving/saving_lib.py

For this format to be fully self contained, it should also have the assets.

hertschuh commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -24,6 +23,13 @@
+    """Get the orbax multihost module with lazy import."""
+    import orbax.checkpoint as ocp
+
+    return ocp.multihost
```
Can we use the "normal" LazyModule approach? Can it piggyback on the existing `keras.src.utils.module_utils.ocp`? Or do we need a separate one?

hertschuh commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -167,6 +181,57 @@ def __init__(
+            # Check if JAX distributed client is initialized
+            # (indicates multihost setup)
+            return multihost.is_jax_distributed_client_initialized()
+        except (ImportError, AttributeError):
```
Is there any case when the  normal orbax checkpoint v1 can be imported but the orbax multihost cannot be imported? I think this can stay in the LazyModule and we don't need to handle it here.

hertschuh commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,150 @@
+        raise ImportError(
+            "Orbax checkpoint support requires the 'orbax-checkpoint' package. "
+            "Install it with: pip install orbax-checkpoint"
+        )
```
If you remove this try / except, you'll get the correct message from the LazyModule, no need to replicate this logic.

amitsrivastava78 replied:
Done

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,150 @@
+            f"Orbax checkpoint at {filepath} does not contain model "
+            "configuration. Cannot recreate model from checkpoint. This "
+            "checkpoint was created with an older version that did not save "
+            "model configuration."
```
Done

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,150 @@
+        # Apply gradients to create optimizer slots
+        model.optimizer.apply_gradients(
+            zip(zero_grads, model.trainable_variables)
+        )
```
Since .keras saves optimizer state only if variables exist, and loading gracefully skips if they don't, no dummy initialization is needed
Why Orbax Needs the Dummy
Orbax checkpoints save complete optimizer state (momentum, velocity, etc.) regardless of whether the optimizer has been built
The set_state_tree() method expects optimizer variables to already exist when setting optimizer_variables
Unlike .keras which skips loading if variables don't exist, Orbax requires the variables to be present for restoration
The dummy apply_gradients with zero gradients creates these slot variables without affecting model weights, enabling proper state restoration

hertschuh replied:
> Orbax checkpoints save complete optimizer state (momentum, velocity, etc.) regardless of whether the optimizer has been built

But if the optimizer has not been built, there are no optimizer state variables (momentums, etc.). Orbax has no idea if there should be optimizer variables or not and how many. So I don't see how this is Orbax specific.

The other thing is that this `apply_gradients` is going to modify the model weights. Just because the gradient is zero doesn't mean that the delta applied by the optimizer is not zero. We absolutely don't want to modify the model weight. There has to be another mechanism in place for the other formats.

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,150 @@
+        return ocp.is_orbax_checkpoint(filepath)
+    except (ImportError, AttributeError):
+        # Fallback to check for orbax.checkpoint file if Orbax API not available
+        return os.path.isfile(os.path.join(filepath, "orbax.checkpoint"))
```
The _is_orbax_checkpoint(filepath) function is designed to work when filepath is the top-level checkpoint directory (the root folder containing the orbax.checkpoint metadata file and step subdirectories).

ocp.is_orbax_checkpoint(filepath) checks for Orbax checkpoint metadata in the given directory
The fallback [os.path.isfile(os.path.join(filepath, "orbax.checkpoint"))]looks for the orbax.checkpoint file directly in [filepath]
If [filepath] is a sub-folder (like a specific step directory, e.g., [checkpoint_dir/0], neither check will work:

The orbax.checkpoint file exists at the root level, not in step subdirs
ocp.is_orbax_checkpoint("checkpoint_dir/0")] would likely return False
The fallback would check for ["checkpoint_dir/0/orbax.checkpoint"] which doesn't exist
In the load_weights function, there's additional logic to detect step subdirectories and handle them, but _is_orbax_checkpoint itself only returns True for the root checkpoint directory. If a step subdir is passed to this function, it would incorrectly return False.

hertschuh reviewed (commented):
I didn't see where the hook is to support `model.load_weights`. Did I miss something?

hertschuh commented on keras/src/saving/saving_lib.py:
```diff
@@ -707,16 +707,20 @@ def _save_state(
-    if not isinstance(weights_store, (H5IOStore, ShardedH5IOStore, NpzIOStore)):
+    if weights_store is not None and not isinstance(
+        weights_store, (H5IOStore, ShardedH5IOStore, NpzIOStore)
+    ):
```
If looks like you were trying to reuse `_save_state` but didn't finish?

hertschuh replied:
Revert this file. These are dead changes at this point.

hertschuh commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -201,17 +260,46 @@ def _save_checkpoint(self, step, logs=None):
+                            )
+                        else:
+                            asset_dict[key] = value
+                    assets_tree[layer.name] = asset_dict
```
Layer names are not necessarily unique, so we can't use this as a key.

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -308,9 +532,65 @@ def load_weights(model, filepath, skip_mismatch=False, **kwargs):
+                                restored_dict[key] = value
+                        else:
+                            restored_dict[key] = value
+                    layer.asset_data = restored_dict
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint.py:
```diff
@@ -201,17 +260,46 @@ def _save_checkpoint(self, step, logs=None):
                 ]
+            # Include assets even for weights-only checkpoints
+            assets_tree = {}
+            for layer in self.model.layers:
```
Key Changes Made
1. Migrated to save_assets API: Updated the OrbaxCheckpoint to use the modern save_assets/load_assets methods instead of the deprecated asset_data approach.
2. Implemented Recursive Asset Collection: Created a _collect_assets_recursive method that:
    * Traverses the entire KerasSaveable hierarchy using _walk_saveable
    * Collects assets from all objects with save_assets methods, not just top-level layers
    * Properly handles lists of KerasSaveable objects (like model.layers)
    * Skips non-KerasSaveable objects to avoid errors
3. Fixed Validation Consistency: Ensured save and load operations use consistent validation logic.
4. Maintained Compatibility: Works for both save_weights_only=True and save_weights_only=False cases.

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -15,6 +15,215 @@
 
 
+def _load_model_from_orbax_checkpoint(
+    filepath, custom_objects=None, compile=True, safe_mode=True
```
Done

amitsrivastava78 commented:
> The assets support won't work the way it is coded right now, see my comments in the code.
> 
> `saving_lib._save_state` shows how it works, and it's pretty specific:
> 
> * the way it recurses in `KerasSaveable`s (not layers)
> * the way it builds a name / path during the recursion
> * the way it calls `save_assets`
> * etc.
> 
> The easiest would be to reuse it while disabling the weights store (which I think you started).
> 
> The complication is that you would need to create a temp folder and then call `save_assets` and then enumerate files, read them, create a sub-key and base64 encode the content.
> 
> And the reverse when reloading...
> 
> Also, for unit tests, you should test with a `StringLookup` or `IntegerLookup` layer to make sure it works end to end.
> 
> Now... honestly, this is neither indispensable nor urgent.
> 
> Let's first submit this PR without model config and without model assets (meaning reloading is only supported via `load_weights`).
> 
> Then let's decide on how to proceed. Because I think if we do want to support this, we need a bit more, like `model.save` should support this format as an option.
> 
> Last thing I thought about. `model.save_weights` should support the Orbax format for consistency.

Thank you for the detailed feedback on the asset support implementation. You're absolutely right that the current approach was flawed and wouldn't work properly.

I've addressed your suggestions by removing asset support entirely from this PR, as you recommended. The implementation now focuses on clean weights-only functionality for Orbax checkpoints.

hertschuh commented:
> Thank you for the detailed feedback on the asset support implementation. You're absolutely right that the current approach was flawed and wouldn't work properly.
> 
> I've addressed your suggestions by removing asset support entirely from this PR, as you recommended. The implementation now focuses on clean weights-only functionality for Orbax checkpoints.

Oh, but you need to remove the config support also. They go together. Either you have just the weights, or you have the model definition, but the model definition is a combination of the config and the assets.

hertschuh commented on keras/src/saving/saving_api.py:
```diff
@@ -1,18 +1,20 @@
-try:
-    import h5py
-except ImportError:
-    h5py = None
```
Undo this

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/saving/saving_api.py:
```diff
@@ -1,18 +1,20 @@
 import os
 import zipfile
 
+import h5py
```
Can you revert this for now? I think it so that `import keras` still works without `h5py` installed, which can be useful. The truth is that the custom handling logic of `h5py` should be replaced by a standard `LazyModule`.

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/saving/orbax_util.py:
```diff
@@ -0,0 +1,137 @@
+from keras.src.utils.module_utils import ocp
+
+
+def _load_model_from_orbax_checkpoint(
```
Remove for now.

hertschuh commented on keras/src/saving/orbax_util.py:
```diff
@@ -0,0 +1,137 @@
+    return model
+
+
+def _is_orbax_checkpoint(filepath):
```
To follow the standard for utils, remove underscore at the beginning: `is_orbax_checkpoint`

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -1,18 +1,20 @@
-try:
-    import h5py
-except ImportError:
-    h5py = None
```
Done

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -1,18 +1,20 @@
+# Import Orbax functions
+_load_model_from_orbax_checkpoint = orbax_util._load_model_from_orbax_checkpoint
+_is_orbax_checkpoint = orbax_util._is_orbax_checkpoint
+_find_latest_orbax_checkpoint = orbax_util._find_latest_orbax_checkpoint
```
Done

amitsrivastava78 commented on keras/src/saving/saving_lib.py:
```diff
@@ -707,16 +707,20 @@ def _save_state(
-    if not isinstance(weights_store, (H5IOStore, ShardedH5IOStore, NpzIOStore)):
+    if weights_store is not None and not isinstance(
+        weights_store, (H5IOStore, ShardedH5IOStore, NpzIOStore)
+    ):
```
Done

amitsrivastava78 commented on keras/src/saving/orbax_util.py:
```diff
@@ -0,0 +1,137 @@
+        return os.path.isfile(os.path.join(filepath, "orbax.checkpoint"))
+
+
+def _find_latest_orbax_checkpoint(checkpoint_dir):
```
Done

hertschuh commented on keras/src/saving/saving_api.py:
```diff
@@ -208,8 +204,9 @@ def load_model(filepath, custom_objects=None, compile=True, safe_mode=True):
-            "legacy H5 format files (`.h5` extension). "
+            "Keras 3 only supports V3 `.keras` files, "
+            "legacy H5 format files (`.h5` extension), and "
+            "Orbax checkpoints. "
```
Undo this too, this message is now incorrect.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +313,371 @@ def test_initial_value_threshold(self):
     @pytest.mark.requires_trainable_backend
-    def test_checkpoint_loading(self):
-        """Test that saved checkpoints can be loaded and weights restored."""
+    def test_checkpoint_loading_comprehensive(
```
This test should be doing `load_weights` to actually test the "loading" per its name.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +313,371 @@ def test_initial_value_threshold(self):
+                self.assertAllClose,
+                original_state_tree["metrics_variables"],
+                loaded_state_tree["metrics_variables"],
+            )
```
Wait, these have nothing to do with metrics, you should always verify those.

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +313,371 @@ def test_initial_value_threshold(self):
+            state_to_set.update(
+                {
+                    "non_trainable_variables": loaded_state[
+                        "non_trainable_variables"
```
This has nothing to do with metrics, it should be line 402

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +313,371 @@ def test_initial_value_threshold(self):
+        # Skip test if there are more than 2 devices, as these tests are
+        # designed for 2-device scenarios and may not work with more devices
+        if len(devices) > 2:
+            self.skipTest(f"Test for 2 devices, but {len(devices)} available")
```
So this test is always skipped, right?

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -136,60 +142,66 @@ def test_save_best_only(self):
+        loaded_weights = new_model.get_weights()
+        for orig, loaded in zip(original_weights, loaded_weights):
+            self.assertTrue(
+                np.allclose(orig, loaded),
```
Use `self.assertAllClose`.

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
-        model.fit(x, y, epochs=1, callbacks=[callback], verbose=0)
-        callback.wait_until_finished()
+        # Train to create checkpoint
+        epochs = 1 if save_on_background else 1
```
???

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+        new_weights_before = new_model.get_weights()
+        for orig, new in zip(original_weights, new_weights_before):
+            self.assertFalse(
+                np.allclose(orig, new),
```
`self.assertNotAllClose`

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+        loaded_weights = new_model.get_weights()
+        for orig, loaded in zip(original_weights, loaded_weights):
+            self.assertTrue(
+                np.allclose(orig, loaded),
```
`self.assertAllClose`

amitsrivastava78 replied:
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+            self.assertAllClose,
+            state_before_loading["metrics_variables"],
+            state_after_loading["metrics_variables"],
         )
```
Remove, metrics are not saved. How is this passing?

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -208,8 +204,9 @@ def load_model(filepath, custom_objects=None, compile=True, safe_mode=True):
-            "legacy H5 format files (`.h5` extension). "
+            "Keras 3 only supports V3 `.keras` files, "
+            "legacy H5 format files (`.h5` extension), and "
+            "Orbax checkpoints. "
```
Done

amitsrivastava78 commented on keras/src/saving/saving_api.py:
```diff
@@ -308,9 +306,61 @@ def load_weights(model, filepath, skip_mismatch=False, **kwargs):
+
+        # Set the weights
+        model.set_weights(weights)
     else:
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
-            {"trainable_variables": loaded_state["trainable_variables"]}
-        )
+        if os.path.exists(checkpoint_dir):
+            shutil.rmtree(checkpoint_dir)
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+        if os.path.exists(checkpoint_dir):
+            shutil.rmtree(checkpoint_dir)
+        os.makedirs(checkpoint_dir, exist_ok=True)
+        shutil.rmtree(checkpoint_dir)  # Clean it again
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+            self.assertAllClose,
+            state_before_loading["non_trainable_variables"],
+            state_after_loading["non_trainable_variables"],
+        )
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+        if os.path.exists(checkpoint_dir):
+            import shutil
+
+            shutil.rmtree(checkpoint_dir)
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+
+            original_distribution = get_distribution()
+        except (ImportError, AttributeError):
+            pass
```
Done

amitsrivastava78 commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,269 @@ def test_initial_value_threshold(self):
+        from keras.src.distribution import set_distribution
+
+        # Check if we have at least 1 device
+        devices = jax.devices()
```
Done

hertschuh commented on keras/src/callbacks/orbax_checkpoint_test.py:
```diff
@@ -293,287 +311,250 @@ def test_initial_value_threshold(self):
+    )
+    def test_distributed_checkpoint_functionality(self):
+        """Test OrbaxCheckpoint with distributed training."""
+        import os
```
Move `import os` to the top
